### PR TITLE
Paged attention transformation

### PR DESCRIPTION
**Pull request by CuriousPanCake**
wip: added the setup part and half of the first matcher

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+//     for (auto& argument : arguments) {
+//         if () {
+//             outputs.push_back(argument);
+//         } else {
+//             outputs.insert(outputs.end(), argument->outputs().begin(), argument->outputs().end());
+//         }
+//     }
+
+//     return outputs;
+// }
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:27]: TODO: discuss if we need this function in C++

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        };
+
+        auto real_k = take_4d(k_current, k_current_reshaped, k_current2);
+        auto real_v = take_4d(v_current, v_current_reshaped, v_current2);
+
+        // is_cpu() check required here.
+        // kv_cache_type required here (CURRENTLY USING STUB)
+        auto kv_cache_type = element::i64;
+        auto k_parameter = std::make_shared<v0::Parameter>(kv_cache_type, PartialShape{-1, -1, -1, -1});
+        auto v_parameter = std::make_shared<v0::Parameter>(kv_cache_type, PartialShape{-1, -1, -1, -1});
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:121]: TODO: The python code has is_cpu() check here. How to do a similar thing in C++?

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        auto q_reshape = std::make_shared<v1::Reshape>(q_transpose, v0::Constant::create(element::i32, Shape{1}, {0, 0, -1}), true); //correct Constant creation? What Shape to use?
+
+        std::shared_ptr<Node> k_transpose_order = kv_transpose_order; // eeeh, is it a right way to assign Constants? Maybe I should clone somehow?
+        if (pattern_map.find(k_order) != pattern_map.end()) { // reapply transpose found in the graph by manipulating of indices of our Transpose
+            k_transpose_order = std::make_shared<v1::Gather>(pattern_map.at(v_order), kv_transpose_order, v0::Constant::create(element::i32, Shape{}, {0}));
+        }
+        auto k_transpose = std::make_shared<v1::Transpose>(real_k, k_transpose_order);
+        auto k_reshape = std::make_shared<v1::Reshape>(k_transpose, v0::Constant::create(element::i32, Shape{1}, {0, 0, -1}), true);
+
+        std::shared_ptr<Node> v_transpose_order = kv_transpose_order; // eeeh, is it a right way to assign Constants? Maybe I should clone somehow?
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:135]: TODO: is it a right way to assign Constants?

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        auto kv_cache_type = element::i64;
+        auto k_parameter = std::make_shared<v0::Parameter>(kv_cache_type, PartialShape{-1, -1, -1, -1});
+        auto v_parameter = std::make_shared<v0::Parameter>(kv_cache_type, PartialShape{-1, -1, -1, -1});
+        kv_parameters.push_back(k_parameter);
+        kv_parameters.push_back(v_parameter);
+        auto kv_transpose_order = v0::Constant::create(element::i32, Shape{} ,{0, 2, 1, 3}); //correct Constant creation? What Shape to use?
+        auto q_transpose = std::make_shared<v1::Transpose>(real_q, kv_transpose_order);
+        auto q_reshape = std::make_shared<v1::Reshape>(q_transpose, v0::Constant::create(element::i32, Shape{1}, {0, 0, -1}), true); //correct Constant creation? What Shape to use?
+
+        std::shared_ptr<Node> k_transpose_order = kv_transpose_order; // eeeh, is it a right way to assign Constants? Maybe I should clone somehow?
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:128]: TODO: is it a right way to assign Constants?

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            v_reshape,
+            k_parameter,
+            v_parameter
+        };
+        params.insert(params.end(), model_remaining_params.begin(), model_remaining_params.end());
+        std::initializer_list<std::shared_ptr<Node>> additional_params = {scale, alibi_slopes, sliding_window};
+        params.insert(params.end(), additional_params.begin(), additional_params.end());
+
+        // Really not sure if I construct correctly because the Python code uses an additional function
+        auto paged_attention = std::make_shared<ov::op::PagedAttentionExtension>(params);
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:173]: TODO: The python code uses an additional arguments_as_outputs() function, but I don't know if we need it here.

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        std::make_shared<pattern::op::Or>(OutputVector{sdpa_mask, pattern::any_input()})
+    });
+
+    ov::matcher_pass_callback callback = [OV_CAPTURE_CPY_AND_THIS, &kv_parameters, &model_remaining_params, &sliding_window, &parameters_to_remove, &assignes_to_remove](ov::pass::pattern::Matcher& m) {
+        std::cout << "____" << matcher_name << "___Matched___" << std::endl;
+        OPENVINO_ASSERT(m.get_pattern_value_map().find(sdpa) != m.get_pattern_value_map().end());
+        const auto& pattern_map = m.get_pattern_map();
+        OPENVINO_ASSERT(pattern_map.find(sdpa) != pattern_map.end());
+        auto real_q = pattern_map.at(q);
+        // Why do have double check here?
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:99]: Why do have double check here?

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        auto real_v = take_4d(v_current, v_current_reshaped, v_current2);
+
+        // is_cpu() check required here.
+        // kv_cache_type required here (CURRENTLY USING STUB)
+        auto kv_cache_type = element::i64;
+        auto k_parameter = std::make_shared<v0::Parameter>(kv_cache_type, PartialShape{-1, -1, -1, -1});
+        auto v_parameter = std::make_shared<v0::Parameter>(kv_cache_type, PartialShape{-1, -1, -1, -1});
+        kv_parameters.push_back(k_parameter);
+        kv_parameters.push_back(v_parameter);
+        auto kv_transpose_order = v0::Constant::create(element::i32, Shape{} ,{0, 2, 1, 3}); //correct Constant creation? What Shape to use?
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:124]: While rewriting I was confused with shapes I need to provide to constants. For example, what shape is it? Shape{} or Shape{1}?

Similar, this applies to constants storing a single value like on line @146. Is it Shape{} or Shape{1}?

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+            }
+        };
+
+        auto real_k = take_4d(k_current, k_current_reshaped, k_current2);
+        auto real_v = take_4d(v_current, v_current_reshaped, v_current2);
+
+        // is_cpu() check required here.
+        // kv_cache_type required here (CURRENTLY USING STUB)
+        auto kv_cache_type = element::i64;
+        auto k_parameter = std::make_shared<v0::Parameter>(kv_cache_type, PartialShape{-1, -1, -1, -1});
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:120]: In the python code, _kv_cache_type_ is provided as function parameter, but what should we use here? I'm creating a temporary stub for now.

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp (hunk 6 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:99]: Remove double check.

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+        std::shared_ptr<Node> alibi_slopes;
+        if (pattern_map.find(alibi) != pattern_map.end()) {
+            std::cout << "alibi slopes applied" << std::endl;
+            alibi_slopes = std::make_shared<v1::Reshape>(pattern_map.at(alibi), v0::Constant::create(element::i32, Shape{1}, {-1}), false); //here {-1} is interesting in Python TODO: discuss
+            if (alibi_slopes->get_element_type() == element::f32) {
+                alibi_slopes = std::make_shared<v0::Convert>(alibi_slopes, element::f32); //todo
+            }
+        } else {
+            alibi_slopes = v0::Constant::create(element::f32, Shape{1}, {}); // correctly created?
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:158]: Why it is like that? To be discussed.

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp (hunk 1 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:27]: Look like no need for it

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp (hunk 4 above)
Comment (slyalin) [reply src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:128]: That's OK, it is just a shared pointer to the node. In the code here and below we are building the correct transposition order that have a few options. There is a default order that is applied when we are migrating data from main graph flow to PA and back. But also there may be additional transposes in some models that we need to mix with the default one. In this case we are building pure-constant sub-expression based on that default order constant node. This subgraph will be folded with CF later. That's a common approach in transformations -- do not compute values in the code, let CF do that for you. We are not modifying this default `kv_transpose_order` so we reuse the same constant node in many places in the graph, no need for cloning.

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp (hunk 3 above)
Comment (slyalin) [reply src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:135]: See above.

File: src/core/src/pass/sdpa_to_paged_attention.cpp [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+#include "transformations/common_optimizations/TotalSequenceLengthPattern.hpp"
+#include "transformations/common_optimizations/PositionIDsReplacer.hpp"
+
+using namespace ov::op;
+
+bool ov::pass::SDPAToPagedAttention::run_on_model(const std::shared_ptr<ov::Model>& model) {
+    RUN_ON_MODEL_SCOPE(SDPAToPagedAttention);
+
+    // писать сюда
+    // yes sir
```
Comment (slyalin) [inline src/core/src/pass/sdpa_to_paged_attention.cpp:24]: It's time to remove it :)
```suggestion
```

File: src/core/src/pass/sdpa_to_paged_attention.cpp [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+        return false;
+    };
+
+    ParameterVector kv_parameters;
+    std::vector<std::shared_ptr<Node>> assignes_to_remove; // not really used
+    ParameterVector parameters_to_remove;
+    ResultVector results_to_remove;
+
+    if (!has_parameter(model, "position_id")) {
+        auto position_ids = std::make_shared<v0::Parameter>(element::i64, PartialShape{-1, -1}); // position_ids SHOULD I SET NAME?
```
Comment (itikhono) [inline src/core/src/pass/sdpa_to_paged_attention.cpp:69]: >> position_ids SHOULD I SET NAME?
it looks so

File: src/core/src/pass/sdpa_to_paged_attention.cpp [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        position_ids->get_output_tensor(0).set_names({position_ids->get_friendly_name()});
+        model->add_parameters({position_ids});
+        std::cout << "CREATED A NEW position_ids PARAMETER" << std::endl;
+    }
+
+    auto position_ids = std::make_shared<Output<Node>>(model->input("position_id")); // Why is it needed?
```
Comment (itikhono) [inline src/core/src/pass/sdpa_to_paged_attention.cpp:75]: It looks like we are trying to find "position_id" input, lines 68-72 add this input if it doesn't exist

File: src/core/src/pass/sdpa_to_paged_attention.cpp (hunk 11 above)
Comment (CuriousPanCake) [reply src/core/src/pass/sdpa_to_paged_attention.cpp:69]: Resolved.

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp (hunk 4 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:128]: Ok, leaving as it is.

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp (hunk 2 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:121]: Fixed.

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp (hunk 1 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:27]: Removed.

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp (hunk 5 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:173]: removed.

File: src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp (hunk 9 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp:153]: @slyalin , can you explain this line in Python code, please
`alibi_slopes = opset13.constant(np.array([], np.float32))`

Why np.array([]) and what should it be in C++ case?

File: src/core/include/openvino/pass/sdpa_to_paged_attention.hpp [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+namespace pass {
+/**
+ * @brief The transformation replaces KV-cache processing part in LLMs by PagedAttention operation.
+ * \ingroup ov_pass_cpp_api
+ */
+class OPENVINO_API SDPAToPagedAttention : public ModelPass {
+public:
+    OPENVINO_RTTI("SDPAToPagedAttention");
+
+    SDPAToPagedAttention();
```
Comment (slyalin) [inline src/core/include/openvino/pass/sdpa_to_paged_attention.hpp:22]: There is no implementation. Fails when library is trying to load.

Comment (itikhono) [reply src/core/include/openvino/pass/sdpa_to_paged_attention.hpp:22]: should it be =default(); ?

File: src/common/transformations/include/transformations/common_optimizations/TotalSequenceLengthPattern.hpp [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+#include "openvino/op/concat.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "openvino/cc/pass/itt.hpp"
+#include "transformations/utils/utils.hpp"
+#include "transformations_visibility.hpp"
+
+namespace ov {
+namespace pass {
+
+class TRANSFORMATIONS_API TotalSequenceLengthPattern;
```
Comment (ilya-lavrenov) [inline src/common/transformations/include/transformations/common_optimizations/TotalSequenceLengthPattern.hpp:18]: export via `TRANSFORMATIONS_API ` is not required

File: src/bindings/python/src/pyopenvino/core/offline_transformations.cpp [hunk 15]
```diff
@@ -127,4 +129,26 @@ void regmodule_offline_transformations(py::module m) {
...
+        [](std::shared_ptr<ov::Model> model) {
+            std::cout << "___CALLING SDPAToPagedAttention___" << std::endl;
+            ov::pass::Manager manager;
+            manager.register_pass<ov::pass::SDPAToPagedAttention>();
+            manager.run_passes(model);
+            std::cout << "___AFTER CALLING SDPAToPagedAttention___" << std::endl;
+        },
+        py::arg("model"));
+
+    m_offline_transformations.def(
```
Comment (itikhono) [inline src/bindings/python/src/pyopenvino/core/offline_transformations.cpp:145]: debug code, I believe it's better to move it to the separate PR if we want to move PrintModel from CPU transformations to common folder and create python binding for it.

File: src/common/transformations/include/transformations/common_optimizations/PrevSequenceLengthPattern.hpp [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        auto seq = pattern::wrap_type<v8::Gather>({kv_past, pattern::any_input(), pattern::any_input()});
+ 
+        ov::matcher_pass_callback callback = [OV_CAPTURE_CPY_AND_THIS](ov::pass::pattern::Matcher& m) {
+        std::cout << "____" << matcher_name << "___Matched___" << std::endl;
+            // TODO: Check that seq has axis that really takes sequence len but not any other dimension -- use symbolics or look at the constant input
+            auto gather = m.get_match_root();
+            auto target_type = gather->get_output_element_type(0);
+            std::shared_ptr<Node> replacement;
+            if (prev_max_seq_len->get_output_element_type(0) != target_type) {
+                std::cout << "Converting " << prev_max_seq_len->get_output_element_type(0) << " of max_context_len to " << target_type << std::endl;
```
Comment (ilya-lavrenov) [inline src/common/transformations/include/transformations/common_optimizations/PrevSequenceLengthPattern.hpp:43]: here it prints `target_type` as `f32`. I wonder that i64 is converted to f32

**Review by ilya-lavrenov — APPROVED** [review]
opt125m and llama2 work

**Comment by itikhono** [comment]
> opt125m and llama2 work

this list of models work

             "hf-internal-testing/tiny-random-BloomForCausalLM",
             "hf-internal-testing/tiny-random-FalconForCausalLM",
             "hf-internal-testing/tiny-random-Starcoder2ForCausalLM",
             "hf-internal-testing/tiny-random-GPTJForCausalLM",
             "hf-internal-testing/tiny-random-StableLmForCausalLM",
             "hf-internal-testing/tiny-random-LlamaForCausalLM",
             "hf-internal-testing/tiny-random-MistralForCausalLM",
             "hf-internal-testing/tiny-random-OPTForCausalLM",
             "hf-internal-testing/tiny-random-PhiForCausalLM",
             "hf-internal-testing/tiny-random-StableLmForCausalLM",
             "facebook/opt-125m"
             
I will re-open this PR and resolve some comments

**Comment by itikhono** [comment]
a new PR has been opened: https://github.com/openvinotoolkit/openvino/pull/24336

File: src/common/transformations/include/transformations/common_optimizations/PrevSequenceLengthPattern.hpp (hunk 16 above)
Comment (itikhono) [reply src/common/transformations/include/transformations/common_optimizations/PrevSequenceLengthPattern.hpp:43]: fixed

File: src/bindings/python/src/pyopenvino/core/offline_transformations.cpp (hunk 15 above)
Comment (itikhono) [reply src/bindings/python/src/pyopenvino/core/offline_transformations.cpp:145]: fixed

File: src/common/transformations/include/transformations/common_optimizations/TotalSequenceLengthPattern.hpp (hunk 14 above)
Comment (itikhono) [reply src/common/transformations/include/transformations/common_optimizations/TotalSequenceLengthPattern.hpp:18]: done
